x/dsmr/storage_test.go

```diff
@@ -369,3 +373,129 @@ func TestRestartSavedChunks(t *testing.T) {
 	storage = restart()
 	confirmChunkStorage(storage)
 }
+
+func TestChunkProducerRateLimiting(t *testing.T) {
+	chunk, err := newChunk(
+		UnsignedChunk[dsmrtest.Tx]{
+			Producer:    testDefaultProducer,
+			Beneficiary: codec.Address{},
+			Expiry:      1,
+			Txs:         []dsmrtest.Tx{{ID: ids.GenerateTestID(), Expiry: 1_000_000}},
+		},
+		[48]byte{},
+		[96]byte{},
+	)
+	require.NoError(t, err)
+	chunkSize := uint64(len(chunk.bytes))
+
+	testCases := []struct {
+		name                 string
+		expiryTimes          []int64
+		newChunkExpiry       int64
+		weightLimit          uint64
+		minTime              int64
+		acceptedChunksExpiry []int64
+		expectedErr          error
+	}{
+		{
+			name:           "success - first",
+			expiryTimes:    []int64{},
+			newChunkExpiry: 50,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "success - after",
+			expiryTimes:    []int64{1},
+			newChunkExpiry: 50,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "success - before",
+			expiryTimes:    []int64{50},
+			newChunkExpiry: 1,
+			weightLimit:    chunkSize * 5,
+			expectedErr:    nil,
+		},
+		{
+			name:           "fail - localized saturated range",
+			expiryTimes:    []int64{0, 50, 99, 150, 500},
+			newChunkExpiry: 75,
+			weightLimit:    chunkSize * 2,
+			expectedErr:    ErrChunkRateLimitSurpassed,
+		},
+		{
+			name:           "fail - localized saturated range with multiple elements",
+			expiryTimes:    []int64{0, 100, 120, 150, 200},
+			newChunkExpiry: 130,
+			weightLimit:    chunkSize * 3,
+			expectedErr:    ErrChunkRateLimitSurpassed,
+		},
+		{
+			name:                 "success - accepted block clear previous limit",
+			expiryTimes:          []int64{0, 50, 100},
+			newChunkExpiry:       150,
+			weightLimit:          chunkSize * 3,
+			expectedErr:          nil,
+			minTime:              10,
+			acceptedChunksExpiry: []int64{50},
+		},
+		{
+			name:                 "success - expired chunks clear previous limit",
+			expiryTimes:          []int64{50, 100, 150},
+			newChunkExpiry:       200,
+			weightLimit:          chunkSize * 3,
+			expectedErr:          nil,
+			minTime:              55,
+			acceptedChunksExpiry: []int64{},
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			require := require.New(t)
+			rules := &ruleFactory{
+				rules: rules{
+					maxProducerChunkWeight: testCase.weightLimit,
+				},
+			}
+			storage, chunks, _, _, _ := createTestStorage(t, testCase.expiryTimes, []int64{}, rules)
+
+			for _, chunk := range chunks {
+				require.NoError(storage.AddLocalChunkWithCert(chunk, nil))
+			}
+
+			if testCase.minTime != 0 {
```

Comment:
Adding this check creates the possibility of a test case where a field is accidentally ignored ie. `minTime = 0`, but there's a non-zero number of accepted chunks, which will be ignored.

Could we remove this possibility by just removing this check and relying on the fact that `SetMin(0, nil)` will be a no-op that should be safe to call.